src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -581,6 +648,15 @@ object ICustomAstVisitor.VisitUnaryExpression(UnaryExpressionAst unaryExpression
         object ICustomAstVisitor.VisitConvertExpression(ConvertExpressionAst convertExpressionAst)
         {
             var type = convertExpressionAst.Type.TypeName.GetReflectionType();
+
+            if (type == typeof(PSObject) && convertExpressionAst.Child is HashtableAst hashtableAst)
```

Comment:
I guess `[psobject]@{a =1}` would be very rare and thus we don't need to differentiate `[pscustomeobject]@{a=1}` from `[psobject]@{a =1}`.

It would be great if you can add a comment about `[PSCustomObject]` actually returns `PSObject` as the reflection type.

Reply:
Good point!
I didn't even consider the `[psobject]@{a =1}` case. I'll revisit, now that I understand what you meant.